mw_validate_json.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/xeipuuv/gojsonschema"
+
+	"github.com/TykTechnologies/tyk/apidef"
+)
+
+type ValidateJSON struct {
+	BaseMiddleware
+}
+
+func (k *ValidateJSON) Name() string {
+	return "ValidateJSON"
+}
+
+func (k *ValidateJSON) EnabledForSpec() bool {
+	for _, v := range k.Spec.VersionData.Versions {
+		if len(v.ExtendedPaths.ValidateJSON) > 0 {
+			return true
+		}
+	}
+
+	return false
+}
+
+// ProcessRequest will run any checks on the request on the way through the system, return an error to have the chain fail
+func (k *ValidateJSON) ProcessRequest(w http.ResponseWriter, r *http.Request, _ interface{}) (error, int) {
+
+	_, versionPaths, _, _ := k.Spec.Version(r)
+	found, meta := k.Spec.CheckSpecMatchesStatus(r, versionPaths, ValidateJSONRequest)
+	if !found {
+		return nil, http.StatusOK
+	}
+
+	vPathMeta := meta.(*apidef.ValidatePathMeta)
+	fmt.Printf("%T: %+v\n", vPathMeta.ValidateWith, vPathMeta.ValidateWith)
+	if vPathMeta.ValidateWith == nil {
+		return errors.New("no schemas to validate against"), http.StatusInternalServerError
+	}
+
+	rCopy := copyRequest(r)
+	bodyBytes, err := ioutil.ReadAll(rCopy.Body)
+	if err != nil {
+		return err, http.StatusInternalServerError
+	}
+	defer rCopy.Body.Close()
+
+	schema := vPathMeta.ValidateWith
+
+	result, err := k.validate(bodyBytes, schema)
+	if err != nil {
+		return err, http.StatusInternalServerError
+	}
+
+	if !result.Valid() {
+		errStr := "payload validation failed"
+		for _, desc := range result.Errors() {
+			errStr = fmt.Sprintf("%s, %s", errStr, desc)
+		}
+
+		if vPathMeta.ValidationErrorResponseCode == 0 {
+			vPathMeta.ValidationErrorResponseCode = http.StatusUnprocessableEntity
+		}
+
+		return errors.New(errStr), vPathMeta.ValidationErrorResponseCode
+	}
+
+	return nil, http.StatusOK
+}
+
+func (k *ValidateJSON) validate(input []byte, schema map[string]interface{}) (*gojsonschema.Result, error) {
+	inputLoader := gojsonschema.NewBytesLoader(input)
+	rulesLoader := gojsonschema.NewGoLoader(schema)
```

Comment:
this is related to loading/decoding of json schema. I feel like validating/parsing and loading of schema is the task for schema-loading process, we don't need to do this when processing every request inside a middleware

Reply:
Since middleware created per API, I think it should be ok to add some field to ValidateJSON middleware and cache this schema.